text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+type on which the policy is defined.
+
+
+Motivation
+==========
+
+The main motivation for this mechanism is to provide a way to define
+*object level security* (equivalent of row level security in SQL) by declaring
+policies that automatically inject the necessary filters based on the session
+context.
+
+The secondary motivation is the ability to trigger additional processing
+whenever an object in question is read or modified.  This is commonly referred
+to as "triggers" in SQL databases and is very useful to implement patterns
+such as audit logs or to automatically generate external change events.
+
+
+Requirements
+============
+
+Although there is no direct dependency, this RFC requires RFC 1010 (globals)
+to be implemented to be practical.  See `Interaction with globals`_ below
+for details.
+
+
+Specification
+=============
+
+This RFC proposes to add a new schema item type -- *Policy* -- that is
+defined in the context of an object type and specifies the rules of rewriting
+queries that refer to the enclosing object type.
+
+CREATE POLICY
+-------------
+
+Define a new query rewrite policy for a given object type.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    {CREATE|ALTER} OBJECT TYPE <type-name> "{"
+        CREATE POLICY
+            ON { SELECT | INSERT | UPDATE | DELETE } [, ... ]
+            [ WHEN <condition> ]
+            <action>
+    "}"
+
+    # where <action> can be one of:
+
+    PERMIT <element-expr>
+
+    RESTRICT TO <element-expr>
+
+    CHECK <element-expr>
+
+    DO AFTER <set-expr>
+
+    DO ON COMMIT <set-expr>
+
+Policies apply to the given list of statement types.  Not all actions are
+valid for all statement types.  See table below for legal combinations.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to a set containing at least one ``true`` value.
+
+``<element-expr>`` is evaluated for every object affected by the statement
+the same way as a ``FILTER`` clause is evaluated, and can use abbreviated
+paths, for example::
+
+    POLICY ON SELECT RESTRICT TO .rating != 'R'
+
+``<set-expr>`` is evaluated for the entire set of objects affected by the
+statement and can refer to the set as ``__result__``, for example::
+
+    POLICY ON UPDATE DO AFTER (
+        FOR obj IN __result__ UNION (
+        INSERT AuditLog {
+            text := 'updated ' ++ obj.name
+        })
+    )
+
+    # or
+
+    POLICY ON INSERT DO AFTER (
+        INSERT EventLog {
+            text := 'inserted ' ++ count(__result__) ++ ' things'
+        }
+    )
+
+The ``PERMIT`` and ``RESTRICT TO`` actions are *filtering* actions and define
+which elements of the object set are visible in a query.  If at least one
+filtering policy is defined for a type, then all elements become invisible by
+default and must be made visible by at least one ``PERMIT`` policy.
```

Comment:
Does this mean that if we want to have use `RESTRICT TO`, we need to always also have a `PERMIT`? What's the rationale there?

I think it would be clearer if the rule is "If at least one `PERMIT` policy is defined for a type"

Reply:
> Does this mean that if we want to have use `RESTRICT TO`, we need to always also have a `PERMIT`

The rationale is security.  We want to err on the side of returning _less_ data than returning _more_ data in an event of a badly written policy, thus, we start with an empty set and gradually "expand" it with "PERMIT".  This is the same design as PostgreSQL RLS and many other ACL systems where "deny all" is the default once ACL is enabled.

Reply:
I'm not totally convinced by the security argument here. There aren't any *useful* rule sets that have differing behavior based on the decision here: any rule set that has a different behavior as a result of my proposal is one that always returned zero objects before. There's not really for something being subtly wrong.

That said, if this is the standard way to define this sort of thing, then all right. But the security framing seems kind of marginal to me.

Reply:
I do agree somewhat that a requirement to have at least one `PERMIT` policy might be surprising at first, but I also think it's weird if the ACL model switches from an allowlist to a denylist and vice-versa based on a presence of a `PERMIT` action, especially when inheritance is involved.

I think it's beneficial to think about the filtering policies as *permissions* rather than *constraints*.  You start with no permissions and then you explicitly grant them.  I think `RESTRICT TO` wouldn't be used too frequently, and the use case would be to carve some objects out of the otherwise permitted set.

Reply:
So the way to have a single `RESTRICT TO` would be to first have a rule that is `PERMIT` all, and then restrict what is not allowed?